python_modules/dagster/dagster/_core/definitions/asset_daemon_context.py

```diff
@@ -500,7 +500,18 @@ def get_auto_materialize_asset_evaluations(
             # over evaluation to any required neighbor key
             if to_materialize_for_asset:
                 for neighbor_key in self.asset_graph.get_required_multi_asset_keys(asset_key):
-                    evaluations_by_key[neighbor_key] = evaluation._replace(asset_key=neighbor_key)
+                    auto_materialize_policy = self.asset_graph.auto_materialize_policies_by_key.get(
+                        neighbor_key
+                    )
+
+                    if auto_materialize_policy is None:
+                        check.failed(f"Expected auto materialize policy on asset {asset_key}")
+
+                    evaluations_by_key[neighbor_key] = evaluation._replace(
+                        asset_key=neighbor_key
+                    )._replace(
+                        rule_snapshots=auto_materialize_policy.rule_snapshots
```

Comment:
I don't love this pattern, but handling this case gracefully is a bit out of scope for this PR